lib/routes/totp.js

```diff
@@ -0,0 +1,237 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const errors = require('../error')
+const isA = require('joi')
+const P = require('../promise')
+const otplib = require('otplib')
+const qrcode = require('qrcode')
+const METRICS_CONTEXT_SCHEMA = require('../metrics/context').schema
+
+module.exports = (log, db, customs, config) => {
+
+  // Default options for TOTP
+  otplib.authenticator.options = {
+    encoding: 'hex',
+    step: config.step
+  }
+
+  // Currently, QR codes are rendered with the highest possible
+  // error correction, which should in theory allow clients to
+  // scan the image better.
+  // Ref: https://github.com/soldair/node-qrcode#error-correction-level
+  const qrCodeOptions = {errorCorrectionLevel: 'H'}
+
+  P.promisify(qrcode.toDataURL)
+
+  return [
+    {
+      method: 'POST',
+      path: '/totp/create',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            metricsContext: METRICS_CONTEXT_SCHEMA
+          }
+        },
+        response: {
+          schema: isA.object({
+            qrCodeUrl: isA.string().required(),
+            secret: isA.string().required()
+          })
+        }
+      },
+      handler(request, reply) {
+        log.begin('totp.create', request)
+
+        let response
+        const sessionToken = request.auth.credentials
+        const uid = sessionToken.uid
+
+        customs.check(request, 'totpCreate')
+          .then(createTotpToken)
+          .then(emitMetrics)
+          .then(createResponse)
+          .then(() => reply(response), reply)
+
+        const secret = otplib.authenticator.generateSecret()
+
+        function createTotpToken() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.createTotpToken(uid, secret, 0)
+        }
+
+        function createResponse() {
+          const otpauth = otplib.authenticator.keyuri(sessionToken.email, config.serviceName, secret)
+
+          return qrcode.toDataURL(otpauth, qrCodeOptions)
+            .then((qrCodeUrl) => {
+              response = {
+                qrCodeUrl,
+                secret
+              }
+            })
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'totpToken.created',
+            uid: uid
+          })
+          return request.emitMetricsEvent('totpToken.created', {uid: uid})
```

Comment:
Can you update `docs/metrics-events.md` with details of all the new `totpToken.created` flow events? And add some local test coverage to make sure they're emitted when they should be?

(apologies if you're seeing this comment twice btw, I swear I submitted it before but I can't see it anywhere on the PR)